Content.Shared/RCD/Systems/RCDSystem.cs

```diff
@@ -35,312 +41,588 @@ public sealed class RCDSystem : EntitySystem
+    private void OnDoAfterAttempt(EntityUid uid, RCDComponent component, DoAfterAttemptEvent<RCDDoAfterEvent> args)
     {
-        // sus client crash why
         if (args.Event?.DoAfter?.Args == null)
             return;
 
-        var location = GetCoordinates(args.Event.Location);
+        // Exit if the RCD prototype has changed
+        if (component.ProtoId != args.Event.StartingProtoId)
+            return;
 
-        var gridId = location.GetGridUid(EntityManager);
-        if (!HasComp<MapGridComponent>(gridId))
-        {
-            location = location.AlignWithClosestGridTile();
-            gridId = location.GetGridUid(EntityManager);
-            // Check if fixing it failed / get final grid ID
-            if (!HasComp<MapGridComponent>(gridId))
-                return;
-        }
+        // Ensure the RCD operation is still valid
+        var location = GetCoordinates(args.Event.Location);
 
-        var mapGrid = Comp<MapGridComponent>(gridId.Value);
-        var tile = mapGrid.GetTileRef(location);
+        if (!TryGetMapGridData(location, out var mapGridData))
+            return;
 
-        if (!IsRCDStillValid(uid, comp, args.Event.User, args.Event.Target, mapGrid, tile, args.Event.StartingMode))
+        if (!IsRCDOperationStillValid(uid, component, mapGridData.Value, args.Event.Target, args.Event.User))
             args.Cancel();
     }
 
-    private void OnDoAfter(EntityUid uid, RCDComponent comp, RCDDoAfterEvent args)
+    private void OnDoAfter(EntityUid uid, RCDComponent component, RCDDoAfterEvent args)
     {
+        if (args.Cancelled && _net.IsServer)
+            QueueDel(EntityManager.GetEntity(args.Effect));
+
         if (args.Handled || args.Cancelled || !_timing.IsFirstTimePredicted)
             return;
 
-        var user = args.User;
+        args.Handled = true;
+
         var location = GetCoordinates(args.Location);
 
-        var gridId = location.GetGridUid(EntityManager);
-        if (!HasComp<MapGridComponent>(gridId))
+        if (!TryGetMapGridData(location, out var mapGridData))
+            return;
+
+        // Ensure the RCD operation is still valid
+        if (!IsRCDOperationStillValid(uid, component, mapGridData.Value, args.Target, args.User))
+            return;
+
+        // Finalize the operation
+        FinalizeRCDOperation(uid, component, mapGridData.Value, args.Target, args.User);
+
+        // Play audio and consume charges
+        _audio.PlayPredicted(component.SuccessSound, uid, args.User);
+        _charges.UseCharges(uid, args.Cost);
+    }
+
+    private void OnRCDconstructionGhostRotationEvent(RCDConstructionGhostRotationEvent ev)
+    {
+        var uid = GetEntity(ev.NetEntity);
+
+        if (!TryComp<RCDComponent>(uid, out var rcd))
+            return;
+
+        rcd.ConstructionDirection = ev.Direction;
+        Dirty(uid, rcd);
+    }
+
+    #endregion
```

Comment:
If you're not relying upon localevents being relayed via ActivatableUI or the likes then you will need to manually handle verifying this otherwise malicious clients can just set whatever RCD's they want.

Reply:
Think I've solved this by having the local player session added to the raised network event. On the server I check that the associated player ent exists, has hands, and is holding the specified RCD in their active hand, before it changes the component